tests/components/xiaomi_miio/test_config_flow.py

```diff
@@ -0,0 +1,138 @@
+        mock_gateway.gateway_info.mac_address = mac_address
+        mock_gateway.gateway_info.hardware_version = hardware_version
+        mock_gateway.gateway_info.firmware_version = firmware_version
+    else:
+        mock_gateway.gateway_info = None
+
+    return mock_gateway
+
+
+async def test_config_flow_step_user_no_device(hass):
+    """Test config flow ."""
+    result = await hass.config_entries.flow.async_init(
+        const.DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "user"
+    assert result["errors"] == {}
+
+    result = await hass.config_entries.flow.async_configure(result["flow_id"], {},)
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "user"
+    assert result["errors"] == {"base": "no_device_selected"}
+
+
+async def test_config_flow_step_gateway_connect_error(hass):
+    """Test config flow ."""
+    result = await hass.config_entries.flow.async_init(
+        const.DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "user"
+    assert result["errors"] == {}
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], {config_flow.CONF_GATEWAY: True},
+    )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "gateway"
+    assert result["errors"] == {}
+
+    mock_gateway = get_mock_gateway(False)
+
+    with patch(
+        "homeassistant.components.xiaomi_miio.config_flow.ConnectXiaomiGateway",
+        return_value=mock_gateway,
+    ), patch(
+        "homeassistant.components.xiaomi_miio.ConnectXiaomiGateway",
+        return_value=mock_gateway,
+    ):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {CONF_HOST: TEST_HOST, CONF_NAME: TEST_NAME, CONF_TOKEN: TEST_TOKEN},
+        )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "gateway"
+    assert result["errors"] == {"base": "connect_error"}
+
+
+async def test_config_flow_gateway_succes(hass):
+    """Test config flow ."""
+    result = await hass.config_entries.flow.async_init(
+        const.DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "user"
+    assert result["errors"] == {}
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], {config_flow.CONF_GATEWAY: True},
+    )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "gateway"
+    assert result["errors"] == {}
+
+    mock_gateway = get_mock_gateway()
+
+    with patch(
+        "homeassistant.components.xiaomi_miio.config_flow.ConnectXiaomiGateway",
+        return_value=mock_gateway,
+    ), patch(
+        "homeassistant.components.xiaomi_miio.ConnectXiaomiGateway",
+        return_value=mock_gateway,
```

Comment:
```suggestion
```

Reply:
Instead we need to patch `async_setup_entry` in the integration module to avoid setting up the integration when creating the entry. See eg:

https://github.com/home-assistant/core/blob/5f775c14278074a75464b538e69a66813a31e549/tests/components/hue/test_config_flow.py#L19-L23